### PR TITLE
Converting Target kind and crate_types to enum representation

### PR DESCRIPTION
**Pull request by michaelciraci**
Converting Target kind and crate_types to enum representation, along with an Unknown(String) as a fallback.

This is in response to the following issue: https://github.com/oli-obk/cargo_metadata/issues/239

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -554,43 +554,158 @@ pub struct Target {
...
+}
+
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+/// Kind of target.
+///
+/// The possible values are `example`, `test`, `bench`, `custom-build` and
+/// [Cargo crate types](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#the-crate-type-field):
+/// `bin`, `lib`, `rlib`, `dylib`, `cdylib`, `staticlib`, `proc-macro`.
+///
+/// Other possible values may be added in the future.
```
Comment (oli-obk) [inline src/lib.rs:604]: Please mark it as non_exhaustive so we can add new variants without it being a breaking change

**Comment by oli-obk** [comment]
Thanks for working on this.

Please update the changelog, too

**Comment by michaelciraci** [comment]
I was looking into why it failed for `nightly`, and it looks like the `main` branch is also failing `nightly`. I have the fix. Do you want me to add it here to get this `nightly` pipeline working again, or open a new merge request?

**Comment by oli-obk** [comment]
Either is fine with me. Whatever works for you

File: tests/test_samples.rs [hunk 2]
```diff
@@ -208,7 +208,7 @@ fn all_the_fields() {
         }
     );
     assert_eq!(meta.workspace_members.len(), 1);
-    assert!(meta.workspace_members[0].to_string().starts_with("all"));
+    assert!(meta.workspace_members[0].to_string().contains("all"));
```
Comment (oli-obk) [inline tests/test_samples.rs:211]: What changed here? Can we get the original display impl back with a manual impl?

Comment (michaelciraci) [reply tests/test_samples.rs:211]: In `stable`, that string is: `all 0.1.0 (path+file:///Users/michaelciraci/Documents/cargo_metadata/tests/all)`
In `nightly`, that string is: `path+file:///Users/michaelciraci/Documents/cargo_metadata/tests/all#0.1.0`

That's why the nightly tests were failing.

**Comment by michaelciraci** [comment]
It looks like the latest version of [cargo-platform](https://crates.io/crates/cargo-platform) (v0.1.6), has a msrv of 1.70. However v0.1.5 works with 1.56. I put an upper limit on `cargo-platform` to keep the msrv of 1.56.

**Comment by jplatte** [comment]
> I put an upper limit on `cargo-platform` to keep the msrv of 1.56.

That is generally a very bad idea, as it leads to un-resolvable dependency trees when another dependency in the tree requires `cargo-platform = "0.1.6"`. Instead, I would recommend updating the MSRV CI job to run `cargo update -p cargo-platform --precise 0.1.5` before checking compilation. There are also other solutions that don't require touching the CI job (or `Cargo.toml`) whenever a dependency bumps its MSRV in a patch release, but they're more involved; maybe the maintainers have a preference towards one such solution, I'm just a contributor.

**Comment by michaelciraci** [comment]
Yeah, there are multiple ways to deal with this and I'll defer to the maintainers on this.

The pull request was really for something else, so I can revert the `cargo-platform` change and that can be dealt with in a different pull request as well.

**Comment by oli-obk** [comment]
> The pull request was really for something else, so I can revert the `cargo-platform` change and that can be dealt with in a different pull request as well.

yes please remove that commit

**Comment by oli-obk** [comment]
Please remove the commits, instead of using a new commit to revert the previous one

**Comment by michaelciraci** [comment]
Should be good now :)

**Comment by oli-obk** [comment]
Thanks!

**Pull request merged**